### ISSUE TITLE
FontFaceSet.ready resolve-time

### ISSUE THREAD

**Issue by gsnedders** [w3c/csswg-drafts#4248]
https://drafts.csswg.org/css-font-loading/#font-face-set-ready

This appears to be pretty vague as to when it resolves:

> The ready attribute contains a Promise which is resolved when the document is done loading fonts, which provides a way for authors to avoid having to keep track of which fonts have or haven’t been loaded before examining content which may be affected by loading fonts.

At what point is the document "done loading fonts"? It would be good to add some cross-reference to something for this. Which document? Or is this meant to be informative and the normative definition that in "switch the FontFaceSet to loaded"?

Per https://github.com/web-platform-tests/wpt/pull/17219#issuecomment-521909714, it seems Blink and Gecko wait until after the first layout, whereas WebKit resolves once it has loaded but before layout is done (should add tests for fallback!). The IDL block has a comment `// async notification that font loading and layout operations are done` but the "layout operations" part doesn't seem to be backed up by any normative text?

**Comment by emilio** [comment]
Getting FontFaceSet.ready _may_ actually trigger layout in Blink and Gecko at least: https://bugzilla.mozilla.org/show_bug.cgi?id=1466004 / https://bugzilla.mozilla.org/show_bug.cgi?id=1471045 / https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/css/font_face_set_document.cc?l=105&rcl=e70e652d516c7d14d50e547aae2da1690c4ae54c.

(Note that triggering layout != triggering RAF callbacks)

**Comment by rniwa** [comment]
Also see https://github.com/w3c/csswg-drafts/issues/1081

**Comment by rniwa** [comment]
Why doesn't this definition refer to the concept of [pending on the environment](https://drafts.csswg.org/css-font-loading/#fontfaceset-pending-on-the-environment)? That's the difference between when this promise is resolved vs. FontFaceSet is no loner pending on the environment.

**Comment by rniwa** [comment]
Looks like Gecko & Blink's implementations are racy with regards to `load` & `DOMContentLoaded` events whereas WebKit's implementation is incorrect (it doesn't wait for the document to finish loading).

**Comment by tabatkins** [comment]
> Or is this meant to be informative and the normative definition that in "switch the FontFaceSet to loaded"?

Yes, this; your quoted text is an informative introduction to the attribute; the algo later in the spec actually manipulates the promise (and does refer to "pending on the environment").

Does anything actually need to be done here, beyond what #1081 is already about?

**Comment by rniwa** [comment]
I think we need to make sure the informative section is noted as such. There is no annotation that it's non-normative so I thought this was the normative text.

**Comment by tabatkins** [comment]
That's reasonable. I'll see if I can gin up some styles that indicate it's a vague description, like the domintro boxes in WHATWG specs.

**Comment by emilio** [comment]
> Looks like Gecko & Blink's implementations are racy with regards to load & DOMContentLoaded events whereas WebKit's implementation is incorrect (it doesn't wait for the document to finish loading).

Can you elaborate? You mean that the promise will be resolved if you add a stylesheet between `DOMContentLoaded` and `load`? I don't think that's the case (though it's subtle).

**Comment by rniwa** [comment]
> Can you elaborate? You mean that the promise will be resolved if you add a stylesheet between `DOMContentLoaded` and `load`? I don't think that's the case (though it's subtle).

In the following example, Firefox would sometimes resolves "ready" before `DOMContentLoaded` is fired.

```html
<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8">
<style>
@font-face {
    font-family: Ahem;
    src: url("./fonts/Ahem.ttf");
}
</style>
<script>
window.addEventListener("load", () => {
    document.getElementById('log').textContent += 'loaded: ' + document.fonts.check('10px Ahem') + '\n';
    document.getElementById('log').textContent += 'document.fonts.size: ' + document.fonts.size + '\n';
});
window.addEventListener("DOMContentLoaded", () => {
    document.getElementById('log').textContent += 'DOMContentLoaded: ' + document.fonts.check('10px Ahem') + '\n';
});
document.fonts.ready.then(() => {
    document.getElementById('log').textContent += 'ready: ' + document.fonts.check('10px Ahem') + '\n';
});
</script>
</head>
<body>
<pre id="log"></pre>
<div id="element1" style="font-size: 10px; font-family: Ahem;">hello</div>
</body>
</html>
```
